packages/roosterjs-content-model-plugins/lib/autoFormat/list/keyboardListTrigger.ts

```diff
@@ -1,65 +1,86 @@
-            const listStyleType = getListTypeStyle(
-                model,
-                shouldSearchForBullet,
-                shouldSearchForNumbering
-            );
-            if (listStyleType) {
-                const segmentsAndParagraphs = getSelectedSegmentsAndParagraphs(model, false);
-                if (segmentsAndParagraphs[0] && segmentsAndParagraphs[0][1]) {
-                    segmentsAndParagraphs[0][1].segments.splice(0, 1);
+        editor.formatContentModel(
+            (model, context) => {
+                const listStyleType = getListTypeStyle(
+                    model,
+                    shouldSearchForBullet,
+                    shouldSearchForNumbering
+                );
+                if (listStyleType) {
+                    const segmentsAndParagraphs = getSelectedSegmentsAndParagraphs(model, false);
+                    if (segmentsAndParagraphs[0] && segmentsAndParagraphs[0][1]) {
+                        segmentsAndParagraphs[0][1].segments.splice(0, 1);
+                    }
+                    const { listType, styleType, index } = listStyleType;
+                    triggerList(model, listType, styleType, index);
+                    context.canUndoByBackspace = true;
+
+                    return true;
                 }
-                const { listType, styleType, index } = listStyleType;
-                triggerList(editor, model, listType, styleType, index);
-                rawEvent.preventDefault();
-                normalizeContentModel(model);
 
-                return true;
+                return false;
+            },
+            {
+                apiName: 'autoToggleList',
             }
-            return false;
-        });
+        );
     }
 }
 
 const triggerList = (
-    editor: IEditor,
     model: ContentModelDocument,
     listType: 'OL' | 'UL',
     styleType: number,
     index?: number
 ) => {
     setListType(model, listType);
     const isOrderedList = listType == 'OL';
-    // If the index < 1, it is a new list, so it will be starting by 1, then no need to set startNumber
-    if (index && index > 1 && isOrderedList) {
-        setListStartNumber(editor, index);
+    const listItem = getFirstSelectedListItem(model);
+    if (listItem) {
+        const listItems = findListItemsInSameThread(model, listItem);
+        const levelIndex = listItem.levels.length - 1;
+        // If the index < 1, it is a new list, so it will be starting by 1, then no need to set startNumber
+        if (index && index > 1 && isOrderedList) {
+            const level = listItem?.levels[levelIndex];
+            if (level) {
+                level.format.startNumberOverride = index;
+            }
+        }
+
+        listItems.forEach(listItem => {
+            const level = listItem.levels[levelIndex];
+            if (level) {
+                updateListMetadata(level, metadata =>
+                    Object.assign(
+                        {},
+                        metadata,
```

Comment:
Are these code copied from the format API? Can we try not duplicate the code?

Reply:
I split in two apis